tests/systemtests/upgrade_test.go

```diff
@@ -0,0 +1,129 @@
+	legacySut.SetExecBinary(legacyBinary)
+	legacySut.SetTestnetInitializer(systest.LegacyInitializerWithBinary(legacyBinary, legacySut))
+	legacySut.SetupChain()
+	v50CLI := systest.NewCLIWrapper(t, legacySut, systest.Verbose)
+	v50CLI.AddKeyFromSeed("account1", testSeed)
+
+	modifications := [][]string{
+		{"genesis", "add-genesis-account", v50CLI.AddKey("foo"), "10000000000stake"},
+		{"genesis", "add-genesis-account", v50CLI.AddKey("bar"), "10000000000stake"},
+		{"genesis", "add-genesis-account", v50CLI.AddKey("baz"), "10000000000stake"},
+	}
+	for _, extraAccount := range extraAccounts {
+		modifications = append(modifications, []string{"genesis", "add-genesis-account", extraAccount.address, extraAccount.balance})
+	}
+
+	legacySut.ModifyGenesisCLI(t,
+		modifications...,
+	)
+
+	return v50CLI, legacySut
+}
+
+func TestChainUpgrade(t *testing.T) {
+	// Scenario:
+	// start a legacy chain with some state
+	// when a chain upgrade proposal is executed
+	// then the chain upgrades successfully
+	systest.Sut.StopChain()
+
+	currentBranchBinary := systest.Sut.ExecBinary()
+	currentInitializer := systest.Sut.TestnetInitializer()
+
+	legacyBinary := systest.WorkDir + "/binaries/v0.50/simd"
+	systest.Sut.SetExecBinary(legacyBinary)
+	systest.Sut.SetTestnetInitializer(systest.NewModifyConfigYamlInitializer(legacyBinary, systest.Sut))
+	systest.Sut.SetupChain()
+
+	votingPeriod := 5 * time.Second // enough time to vote
+	systest.Sut.ModifyGenesisJSON(t, systest.SetGovVotingPeriod(t, votingPeriod))
+
+	systest.Sut.StartChain(t, fmt.Sprintf("--halt-height=%d", upgradeHeight+1))
+
+	cli := systest.NewCLIWrapper(t, systest.Sut, systest.Verbose)
+	govAddr := sdk.AccAddress(address.Module("gov")).String()
+	// submit upgrade proposal
+	proposal := fmt.Sprintf(`
+{
+ "messages": [
+  {
+   "@type": "/cosmos.upgrade.v1beta1.MsgSoftwareUpgrade",
+   "authority": %q,
+   "plan": {
+    "name": %q,
+    "height": "%d"
+   }
+  }
+ ],
+ "metadata": "ipfs://CID",
+ "deposit": "100000000stake",
+ "title": "my upgrade",
+ "summary": "testing"
+}`, govAddr, upgradeName, upgradeHeight)
+	proposalID := cli.SubmitAndVoteGovProposal(proposal)
+	t.Logf("current_height: %d\n", systest.Sut.CurrentHeight())
+	raw := cli.CustomQuery("q", "gov", "proposal", proposalID)
+	t.Log(raw)
+
+	systest.Sut.AwaitBlockHeight(t, upgradeHeight-1, 60*time.Second)
+	t.Logf("current_height: %d\n", systest.Sut.CurrentHeight())
+	raw = cli.CustomQuery("q", "gov", "proposal", proposalID)
+	proposalStatus := gjson.Get(raw, "proposal.status").String()
+	require.Equal(t, "PROPOSAL_STATUS_PASSED", proposalStatus, raw)
+
+	t.Log("waiting for upgrade info")
+	systest.Sut.AwaitUpgradeInfo(t)
+	systest.Sut.StopChain()
+
+	t.Log("Upgrade height was reached. Upgrading chain")
+	systest.Sut.SetExecBinary(currentBranchBinary)
+	systest.Sut.SetTestnetInitializer(currentInitializer)
+	systest.Sut.StartChain(t)
```

Comment:
can we assert that the correct height was reached? and that the new chain starts at upgradeHeight + 1?

Reply:
added assertion [assert block height](https://github.com/cosmos/cosmos-sdk/pull/24182/commits/0c55c097df74819f8bfd0f2138515d7f381cd971)